GraphQLDataProvider: Always fetch current user's data

JS-BC opened the pull request:
This PR introduces a change to the GraphQLDataProvider to update the cached local metadata's current user info with the latest information from the server. This change addresses two issues:

1 - Whenever a user's set of roles have been changed, the changes are not reflected after a page refresh
2 - Logging out does not clear the local metadata, as such its possible to be logged in as user A but using user B to make calls to the API

AN-BC commented on packages/GraphQLDataProvider/src/graphQLDataProvider.ts:
```diff
@@ -57,6 +57,13 @@ export class GraphQLProviderConfigData extends ProviderConfigDataBase {
+     * The Email of the current user. This field is used during the bootstrap process to determine if the local metadata should be refreshed.
+     * If the email is different than the user's email in the local metadata, then the metadata will be refreshed from the server
+     */
+    get UserEmail(): string { return this.Data.UserEmail }
```
@JS-BC  are you working on this, I don't see any changes

JS-BC commented:
@AN-BC Made a few changes:

- `CurrentUserEmail` field was moved up to the base provider class, and removed from the GraphQLDataProvider and SQLServerDataProvider classes
- Added a new `options` optional parameter to the GraphQLDataProvider and SQLSeverDataProvider constructor, allowing us in the future to populate the base class' `data` property without having to make changes to the constructor or elsewhere in the code
- Added descriptions to the SQLServerDataProvider constructor parameters

AN-BC commented on packages/MJCore/src/generic/interfaces.ts:
```diff
@@ -15,20 +15,29 @@ export class ProviderConfigDataBase {
-    constructor(data: any, MJCoreScemaName: string = '__mj', includeSchemas?: string[], excludeSchemas?: string[]) {
+
+    public get CurrentUserEmail(): string {
+        return this._data.CurrentUserEmail;
```
@JS-BC  what I don't understand ---- the `data` property is there to support sub-classes having extensions to the core data elements the base class supports. If we want `CurrentUserEmail` to exist in the base class, why not just have a string property? Also, JSDoc is needed for the getter property.

AN-BC commented on packages/SQLServerDataProvider/src/SQLServerDataProvider.ts:
```diff
@@ -84,34 +84,46 @@ import { EntityActionEngineServer } from '@memberjunction/actions';
+   */
   constructor(
     dataSource: any,
     currentUserEmail: string,
```
@JS-BC  did you add `currentUserEmail` to this constructor as a required second position param? If so, all users of this class will break as it wasn't there before, I don't believe? Or was it?

JS-BC replied:
@AN-BC No, currentUserEmail has always been a second parameter for the sql server data provider

AN-BC commented on packages/GraphQLDataProvider/src/graphQLDataProvider.ts:
```diff
@@ -81,19 +82,22 @@ export class GraphQLProviderConfigData extends ProviderConfigDataBase {
+            data = {...data, ...options};
+        }
+        
+        super(data, MJCoreSchemaName, includeSchemas, excludeSchemas);
```
@JS-BC  GraphQLDataProvider isn't using the current user email property and passing it along to the base constructor???

JS-BC replied:
@AN-BC No, it never has. But with the new options param, im thinking you can set it that way

AN-BC replied:
@JS-BC seems like this got dropped, please address above question.

JS-BC replied:
@AN-BC 
In the GraphQLDataProvider, Ive overrode the IsMetaDataObsolete method to include the check to see if the email of the user passed in matches the email found in the local metadata. That logic is more appropriate at the graph ql level than at the base class level.

I've also removed the currentUserEmail field from the base provider config class, as that field doesnt _really_ need to exist or be exposed at the base class level

These changes arent meant to interact with the ContextCurrentUser, theyre made to ensure that the GraphQLDataProvider refreshes it local metadata in the event the signed in user is different than the user thats in the cache. If we dont want to fetch the signed in user's info every time the provider initializes, then the second best idea (in my opinion) would be to pass the user's email to the provider and use it when we check to see if the local metadata is outdated. 

I had to update the GraphQLDataProvider's constructor in order to pass in the user's email, and from that point I figured both providers would benefit from having a generic, optional Options parameter to support future properties we may want to pass in.

JS-BC commented:
@AN-BC Closing this for now, as I think more about what we talked about last week
